rust/src/interactive/mod.rs

```diff
@@ -0,0 +1,253 @@
+
+pub trait IntoPolyQueryable {
+    fn into_poly(self) -> PolyQueryable;
+}
+
+impl<Q: 'static, A: 'static> IntoPolyQueryable for Queryable<Q, A> {
+    fn into_poly(mut self) -> PolyQueryable {
+        Queryable::new_raw(move |_self: &PolyQueryable, query: Query<dyn Any>| {
+            Ok(match query {
+                Query::External(q) => {
+                    let answer = self.eval(q.downcast_ref::<Q>().ok_or_else(|| {
+                        err!(FailedCast, "query must be of type {}", type_name::<Q>())
+                    })?)?;
+                    Answer::External(Box::new(answer))
+                }
+                Query::Internal(q) => Answer::Internal(self.eval_internal(q)?),
+            })
+        })
+    }
+}
+
+// The previous impl over all Q has an implicit `Sized` trait bound, whereas this parameterizes Q as dyn Any, which is not Sized.
+// Therefore, the compiler recognizes these impls as disjoint.
+impl IntoPolyQueryable for PolyQueryable {
+    fn into_poly(self) -> PolyQueryable {
+        // if already a PolyQueryable, no need to do anything.
+        self
+    }
+}
+
+pub trait FromPolyQueryable {
+    fn from_poly(v: PolyQueryable) -> Self;
+}
+
+impl<Q: 'static, A: 'static> FromPolyQueryable for Queryable<Q, A> {
+    fn from_poly(mut self_: PolyQueryable) -> Self {
+        Queryable::new_raw(move |_self: &Queryable<Q, A>, query: Query<Q>| {
+            Ok(match query {
+                Query::External(query) => {
+                    let answer = self_.eval(query)?;
+
+                    let answer = *answer.downcast::<A>().map_err(|_| {
+                        err!(FailedCast, "failed to downcast to {:?}", type_name::<A>())
+                    })?;
+                    Answer::External(answer)
+                }
+                Query::Internal(q) => Answer::Internal(self_.eval_internal(q)?),
+            })
+        })
+    }
+}
+
+// The previous impl over all Q has an implicit `Sized` trait bound, whereas this parameterizes Q as dyn Any, which is not Sized.
+// Therefore, the compiler recognizes these impls as disjoint.
+impl FromPolyQueryable for PolyQueryable {
+    fn from_poly(self_: Self) -> Self {
+        // if already a PolyQueryable, no need to do anything.
+        self_
+    }
+}
+
+impl<Q: ?Sized> Queryable<Q, Box<dyn Any>> {
+    /// Evaluates a polymorphic query and downcasts to the given type.
+    pub fn eval_poly<A: 'static>(&mut self, query: &Q) -> Fallible<A> {
+        self.eval(query)?
+            .downcast()
+            .map_err(|_| {
+                err!(
+                    FailedCast,
+                    "eval_poly failed to downcast to {}",
+                    std::any::type_name::<A>()
+                )
+            })
+            .map(|b| *b)
+    }
+}
+
+impl PolyQueryable {
+    pub fn into_downcast<Q: ?Sized, A>(self) -> Queryable<Q, A>
+    where
+        Queryable<Q, A>: FromPolyQueryable,
+    {
+        Queryable::<Q, A>::from_poly(self)
+    }
+}
```

Comment:
This doesn't seem to be used in any upcoming branches. Is it intended for user code?

Reply:
It is used for Queryable::new. It could be used for downcasting poly queryables, but people will usually have `Box<dyn Any>` and downcast directly to `Queryable<Q, A>`. Removed.